### PR TITLE
Reduce memory usage in getMultiScaleCosAffinityMatrix function

### PR DESCRIPTION
**Pull request by gabitza-tech**
Reduce memory usage in getMultiScaleCosAffinityMatrix function

Signed-off-by: gabitza-tech <gabriel.pirlogeanu@gmail.com>

# What does this PR do ?

Replace some tensor multiplication in the function for multi scale cos affinity matrix calculation with a single tensor in which we add the repeated embeddings multiplied with its scale weight.

Results show a decrease from 11GB to 2GB for 1 hour audio. It even managed to reduce from above 60GB memory usage for a [53000,53000] tensor to 28GB.
**Collection**: ASR (collections/asr/parts/utils/offline_clustering.py)

# Changelog 
`multiscale_weights = torch.squeeze(multiscale_weights,dim=0).to(device)`
The tensor was [1,6] shaped and had to squeeze the first dimension.
` fused_sim_d = torch.zeros(len(timestamps_in_scales[-1]),len(timestamps_in_scales[-1]))`
Initialized a tensor with the shape [s,s] with 0 values where s is the number of timestamps for the smallest scale. We will add the tensor for each scale to this tensor
` with torch.no_grad():  `
Gradients are not needed for inference.
`fused_sim_d = fused_sim_d + multiscale_weights[scale_idx] * repeated_tensor_1`
Instead of adding the tensor for each scale to a list, then stacking the tensor list and then matrix multiplying it with the weights, we simply add the repeated tensor to the previously initialized tensor before the for loop.

@tango4j is adding followings:
- Clear memory outside torch.jit.script functions since jit-script does not support torch.cuda.empty_cache()
- Fixed the bug where eigen value decomposition does not use GPU. Now GPU usage hits 100%
- Removed torch.jit.script decorators since it slows down if we run without jit.script export.
- Performance (without torch.jit.script) 

sess len in secs | Num of sessions | total time in secs | sec/Iter. | RTFx |  Peak Memory Usage
-- | -- | -- | -- | -- | --
1800 | 10 | 34 | 3.48 | 517.24138 | 3G
3600 | 10 | 70 | 7.02 | 512.82051 | 8G
7200 | 10 | 319 | 31.98 | 225.14071 | 20G

Now, 2hours of audio usually takes 30~40 seconds.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. I would like to ask @titu1994, @redoctopus, @jbalam-nv, or @okuchaiev to review this pull request.
# Additional Information
* Related to https://github.com/NVIDIA/NeMo/issues/5637

### CODE REVIEW

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 1]
```diff
@@ -486,7 +486,7 @@ def getMultiScaleCosAffinityMatrix(
     timestamps_in_scales: List[torch.Tensor],
     device: torch.device = torch.device('cpu'),
 ) -> torch.Tensor:
-    """
+      """
```
Comment (tango4j) [inline nemo/collections/asr/parts/utils/offline_clustering.py:489]: Did you make this tab intentionally? seems like it is causing style error in the test.

File: nemo/collections/asr/parts/utils/offline_clustering.py [hunk 2]
```diff
@@ -506,20 +506,22 @@ def getMultiScaleCosAffinityMatrix(
...
+
+    for scale_idx in scale_list:       
+        with torch.no_grad():            
+            mapping_argmat = session_scale_mapping_list[scale_idx]
+            emb_t = embeddings_in_scales[scale_idx].half().to(device)
+            score_mat_torch = getCosAffinityMatrix(emb_t)
+            repeat_list = getRepeatedList(mapping_argmat, torch.tensor(score_mat_torch.shape[0])).to(device)
+            repeated_tensor_0 = torch.repeat_interleave(score_mat_torch, repeats=repeat_list, dim=0)            
+            repeated_tensor_1 = torch.repeat_interleave(repeated_tensor_0, repeats=repeat_list, dim=1)
+            fused_sim_d = fused_sim_d + multiscale_weights[scale_idx] * repeated_tensor_1
```
Comment (tango4j) [inline nemo/collections/asr/parts/utils/offline_clustering.py:524]: Thank you for accepting my suggestion and pushing this.
Wouldn't it be better to clear memory here by doing the following:

```python
del emb_t, repeated_tensor_0, repeated_tensor_1
torch.empty_cache()
```
I need to check whether this helps or not.

Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py]: This doesn't work in torch.jit.script decorator. I have added torch.empty.cache() outside the clustering code and now it is clearning memory.

File: nemo/collections/asr/parts/utils/offline_clustering.py (hunk 1 above)
Comment (tango4j) [reply nemo/collections/asr/parts/utils/offline_clustering.py]: Fixed this on my own.

**Review by tango4j — APPROVED** [review]

**Comment by tango4j** [comment]
This is passing the test and seems like working well. I will merge this as soon as the newly rebased one passes.

**Review by tango4j — APPROVED** [review]
The suggested changes have been reflected.

**Comment by gabitza-tech** [comment]
@tango4j I wanted to ask if removing the torch.jit.script leads to the same performances in terms of memory usage as before?

**Comment by donishi** [comment]
Also @tango4j, could you please share the configuration you used in the table you added to the PR description? I'm running out of GPU memory (on an RTX3090) when trying this PR with an audio two hours long. This happens with the clustering diarizer + 'general' configuration as well as with the neural diarizer.

**Comment by yoelcabo** [comment]
hey @donishi,

I managed to run a 2 hours and a half audio on a RTX4000 (8GB) with this config:

```
  diarizer.speaker_embeddings.parameters.window_length_in_sec="[3.0,1.5,1.0]" \
  diarizer.speaker_embeddings.parameters.shift_length_in_sec="[1.5,0.75,0.5]" \
  diarizer.speaker_embeddings.parameters.multiscale_weights="[0.33,0.33,0.33]" \
```

With smaller `shift_length_in_sec` and `window_length_in_sec` I was running out of memory. So on top of reducing the number of scales, reducing the amount of windows also helps.

**Comment by donishi** [comment]
That makes sense @yoelcabo, thanks for sharing!

For some reason that setting uses much more than 8GB for me, but it certainly uses less memory than the default window lengths.

**Comment by yoelcabo** [comment]
One other thing I changed and forgot to mention is the batch size:

```
  batch_size=32 \
```

Otherwise, I run out of memory when extracting the embeddings.

**Comment by tango4j** [comment]
@gabitza-tech Correct, removing torch.jit.script does not affect memory allocations
@donishi Each user has differnt GPUs so it's meaningless to share my script. as @yoelcabo described, try with higher base scale parameter (1.0 window, 0.5 hoplength).

Currently, I am also testing low-memory low-speed batch mode for clustering since so many users are raising this issue. I hope we can have clustering system that is also based on batch_size to control the memory usage. 

The memory allocation happens when embedding vectors in  `embedding_in_scales` are loaded to GPU memory.

**Pull request merged**